File: NumbersGameOtus/ClientApp/src/app/services/game.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

interface GameResponse {
  numbers: number[];
  correctNumber: number;
}

export enum GameResult {
  Win = 1,
  Lose = 0
}

@Injectable({
  providedIn: 'root'
})
export class GameService {

  private apiUrl = 'https://localhost:7247/api/game';

  constructor(private http: HttpClient) {}

  startGame(): Observable<GameResponse> {
    return this.http.get<GameResponse>(`${this.apiUrl}/start`);
  }

  playGame(guessedNumber: number, correctNumber: number): Observable<GameResult> {
    const playRequest = {
      guessedNumber,
      correctNumber
    };
    return this.http.post<GameResult>(`${this.apiUrl}/play`, playRequest);
  }
}

File: NumbersGameOtus/Services/GameService.cs
using NumbersGameOtus.Models;

namespace NumbersGameOtus.Services;

public class GameService : IGameService
{
    private readonly Random _random = new();

    public GameResponse StartGame()
    {
        var correctNumber = _random.Next(1, 101);
        var numbers = new HashSet<int> { correctNumber };

        while (numbers.Count < 4)
        {
            numbers.Add(_random.Next(1, 101)); 
        }

        var shuffledNumbers = new List<int>(numbers);
        shuffledNumbers.Sort();  

        return new GameResponse
        {
            Numbers = shuffledNumbers,
            CorrectNumber = correctNumber
        };
    }

    public GameResult PlayGame(int guessedNumber, int correctNumber)
    {
        if (guessedNumber == correctNumber)
        {
            return GameResult.Win;
        }
        else
        {
            return GameResult.Lose;
        }
    }
}